---
title: "EXP_00007_HAR_caMPRA_Tag_Seq_Analysis_Mutagenesis_Analysis_3_Masked_variant_effects"
output:
  html_document:
    df_print: paged
---

# Rationale for Final caMPRA Tag seq analysis
This analysis is to further polish the analysis based on previous preliminary analyses.

There were issues with the first two analyses - the first, while fine, was performed on 1 lane of sequencing for the tag-seq so it wasn't the complete dataset. The second analysis using dnaclust has issues with consistency. The third analysis was performed using umi_tools, but on randomly shuffled (but pairs maintained) subsets. However, there seemed to a filtering issue with the variants that were called, so this analysis is repeating the mutagenesis analysis with new filtered variants (~90K variants vs ~80K variants)

Note that for the Rmd file, will mainly save figures and visualize plots from file (using knitr::include_graphics to save on memory and make sure the plots are available outside of these the notebook)

# outline of overall analysis
1. QC of barcodes and filtering
2. Evaluating the impact of variants on expression
3. Improved masked evaluation of variants (this document)
4. Relationship between variant annotations and effect on expression

# load library
```{r, load library, echo=FALSE}
library(tidyverse)
library(ggpubr)
library(stringdist)
library(GGally)
library(tidygenomics)
library(patchwork)
library(RVenn)
library(superheat)
library(MutationalPatterns)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
library(ref_genome, character.only = TRUE)
library(rstatix)
library(RColorBrewer)
library(AnnotationHub)
library(MafDb.gnomAD.r2.1.hs37d5)
```

# load necessary data
```{r, load data, eval = FALSE}
variant_assigned_barcodes_AF <- read_tsv("../Saved_Data/Mutagenesis_Analysis/variant_assigned_barcodes_AF.txt")
HAR_7x_mut_pD2_filtered_barcodes <- read_tsv("../Saved_Data/Mutagenesis_Analysis/HAR_7x_mut_pD2_filtered_barcodes.txt")
HAR_7x_mut_pD3_filtered_barcodes <- read_tsv("../Saved_Data/Mutagenesis_Analysis/HAR_7x_mut_pD3_filtered_barcodes.txt")
HAR_probe_names <- read_tsv("../Saved_Data/Mutagenesis_Analysis/HAR_probe_names.txt")
```



# Check to see variants are from sample
The caMPRA were captured on GIAB samples, so will compare the variants we have with the variants expected in GIAB samples
GIAB sample: NA12878

```{r, GIAB variant check, eval = FALSE}
HAR_caMPRA_NA12878_GIAB_variants <- read_vcfs_as_granges("/Users/tshin/CA Walsh Lab Dropbox/Taehwan Shin/Walsh Lab/Scripts/Projects/caMPRA_development/EXP_00007_Barcode_Reg_Association/Data/HAR_caMPRA_HG001_GIAB_variants.vcf", sample_names = "GIAB_12878", genome = ref_genome, predefined_dbs_mbs = TRUE) %>%
  as_tibble() %>%
  dplyr::mutate(ALT = unlist(as(ALT, "CharacterList"))) 

# check for overlap
HAR_caMPRA_GIAB_overlap <- inner_join(x = variant_assigned_barcodes_AF,
                                     y = HAR_caMPRA_NA12878_GIAB_variants %>%
                                       dplyr::mutate(CHROM = str_replace(seqnames, pattern = "chr", replacement = "")),
                                     by = c("CHROM" = "CHROM", "POS" = "start", "REF" = "REF", "ALT" = "ALT"))

# number of unique variants: 9031
length(unique(variant_assigned_barcodes_AF$var_id))

# number of GIAB variants that are present in mutagenesis variant set: 834
length(unique(HAR_caMPRA_GIAB_overlap$var_id))
```

So it looks like out of the 9031 variants generated by the random mutagenesis, 834 were already present in the sample (NA12878). In terms of barcode presence, they actually make up half, which makes sense as they will almost always be in the probe that covers the variant.


# redefine control and variant probes
Since it looks like some variants are actually the control sequence, we should move them from the variant probes to control probes, only if the the GIAB variant is the only variant that is there.

Also, would be good to note the variant as GIAB for between region comparisons, but masked for control - variant comparisons (if they are present in both control and variant probes)

since variant information is encoded as variant - barcode with multiple variants per barcode sometimes, we can remove the variants that are GIAB present/(mask them) and see how the control and variant probes look after that
```{r, redefine control and variant probes, eval = FALSE}
# masked random mutagenesis variants
HAR_caMPRA_variants_GIAB_masked <- anti_join(x = variant_assigned_barcodes_AF,
                                     y = HAR_caMPRA_NA12878_GIAB_variants %>%
                                       dplyr::mutate(CHROM = str_replace(seqnames, pattern = "chr", replacement = "")),
                                     by = c("CHROM" = "CHROM", "POS" = "start", "REF" = "REF", "ALT" = "ALT"))

# 75628 barcodes
length(unique(variant_assigned_barcodes_AF$barcode))

# 30800 barcodes
length(unique(HAR_caMPRA_variants_GIAB_masked$barcode))

# 8197 unique variants
length(unique(HAR_caMPRA_variants_GIAB_masked$var_id))
```
We lose quite a few barcodes from the variant pile, but that will be moved to the control pile

```{r, masked control probes pD2, eval = FALSE}
HAR_tag_seq_barcodes_wo_variants_masked_pD2 <- HAR_7x_mut_pD2_filtered_barcodes %>%
  filter(!barcode %in% HAR_caMPRA_variants_GIAB_masked$barcode) %>%
  select(HAR_7x_mut_pD2_TPM, 
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_5,
         probe) %>%
  dplyr::rename(
         HAR_7x_mut.pDonor2.rep_1 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut.pDonor2.rep_2 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut.pDonor2.rep_3 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut.pDonor2.rep_4 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut.pDonor2.rep_5 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_5
         ) %>%
  pivot_longer(-c(HAR_7x_mut_pD2_TPM, probe), names_to = "condition", values_to = "expression") %>%
  separate(col = condition, sep = "\\.", into = c("mut_condition", "promoter", "rep"))

# updated to use rstatix for stats (better record keeping for numbers analyzed)
HAR_tag_seq_barcodes_wo_variants_masked_pD2_stats <- HAR_tag_seq_barcodes_wo_variants_masked_pD2 %>% 
  group_by(probe) %>%
  rstatix::get_summary_stats(expression)

write_tsv(HAR_tag_seq_barcodes_wo_variants_masked_pD2_stats, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_wo_variants_masked_pD2_stats.txt")

# unique probes: 1867 probes
# length(unique(HAR_tag_seq_barcodes_wo_variants_pD2$probe))

# 2082 probes
length(unique(HAR_tag_seq_barcodes_wo_variants_masked_pD2$probe))

# barcodes with variants after GIAB masking
HAR_tag_seq_barcodes_w_variants_masked_pD2 <- HAR_7x_mut_pD2_filtered_barcodes %>%
  filter(barcode %in% HAR_caMPRA_variants_GIAB_masked$barcode) %>%
  select(HAR_7x_mut_pD2_TPM, 
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_5,
         probe,
         barcode) %>%
  dplyr::rename(
         HAR_7x_mut.pDonor2.rep_1 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut.pDonor2.rep_2 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut.pDonor2.rep_3 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut.pDonor2.rep_4 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut.pDonor2.rep_5 = HAR_7x_mut_pD2_cDNA_tag_TPM_norm_log_5
         ) %>%
  pivot_longer(-c(HAR_7x_mut_pD2_TPM, probe, barcode), names_to = "condition", values_to = "expression") %>%
  separate(col = condition, sep = "\\.", into = c("mut_condition", "promoter", "rep"))

HAR_tag_seq_barcodes_w_variants_masked_pD2_stats <- HAR_tag_seq_barcodes_w_variants_masked_pD2 %>% 
  dplyr::group_by(probe, barcode) %>%
  rstatix::get_summary_stats(expression)

write_tsv(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats.txt")

# number of probes that have any variant in them # 1474
# length(unique(HAR_tag_seq_barcodes_w_variants_pD2$probe))

# 1280
length(unique(HAR_tag_seq_barcodes_w_variants_masked_pD2$probe))

# overlap with non-variant containing elements - 869
#sum(unique(HAR_tag_seq_barcodes_w_variants_pD2$probe) %in% unique(HAR_tag_seq_barcodes_wo_variants_pD2$probe))

# overlap after masking procedure - 890/1280, from 869/1474 (with increase in control probes from 1867 to 2082)
sum(unique(HAR_tag_seq_barcodes_w_variants_masked_pD2$probe) %in% unique(HAR_tag_seq_barcodes_wo_variants_masked_pD2$probe))
```
So it appears that the masking procedure removes some barcodes that had variants in them, since those barcodes only had GIAB variants that were masked for the comparison. We have more probes that do not have variants, but for comparison purposes, it is only the probes that have variants that matter


```{r, add variant information, eval = FALSE}
# add variant information - count number of variants per barcode
HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars <- left_join(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats, HAR_caMPRA_variants_GIAB_masked, by = c("barcode" = "barcode")) %>%
  group_by(barcode) %>%
  summarise(variant_count_per_barcode = n())
  
# save information
write_tsv(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars.txt")

HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_plot <- HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars %>%
  ggplot() +
  geom_histogram(aes(x = variant_count_per_barcode), binwidth = 1) +
  theme_pubr()

# plot of variant number per probe as a stacked bargraph
HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked <- HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars %>%
  mutate(total =  n()) %>%
  group_by(variant_count_per_barcode) %>%
  summarise(number_of_probes = n(),
            total = total) %>%
  distinct() %>%
  dplyr::mutate(percentage_of_probes = number_of_probes/total * 100)

HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot <- HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked %>%
  ggplot(.) +
  geom_col(aes(x = "number of variants per probe", y = percentage_of_probes, fill = factor(variant_count_per_barcode))) +
  theme_pubr() +
  geom_text(data = . %>%
              filter(variant_count_per_barcode < 4),
            aes(x = "number of variants per probe", y = percentage_of_probes + 11, label = paste(number_of_probes, "\n(", format(percentage_of_probes, digits = 2, nsmall = 1), "%)")), vjust = 1.5, colour = "white")

ggsave(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_plot.png")
ggsave(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot.png", height = 6, width = 3)
ggsave(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot_wide.png", height = 6, width = 6)

ggsave(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_plot.pdf")
ggsave(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot.pdf", height = 6, width = 3)
ggsave(HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot_wide.pdf", height = 6, width = 6)
```
```{r}
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_plot.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD2_stats_plus_vars_stacked_plot_wide.png")
```

So, it looks like most of the fragments only contain 1 variant - also note that the figures have been updated so barcodes that were filtered out because of low count are also removed prior to normalization

```{r, HAR pD2 control v variant comparison, eval = FALSE}
# perform two sided paired t-tests to identify significantly different log2(mRNA/DNA) ratio between no variant sequences and variant sequences

# prepare overlapping probes (variant containing probes connected to control probes)
# updated code to prevent annoying duplication - however results are very similar as duplications were removed downstream in the previous version (just more messy)
HAR_tag_seq_barcodes_control_v_variant_masked_pD2 <- inner_join(
  HAR_tag_seq_barcodes_w_variants_masked_pD2, 
  HAR_tag_seq_barcodes_wo_variants_masked_pD2, 
  by = c("probe" = "probe", 
         "rep" = "rep",
         "mut_condition" = "mut_condition",
         "promoter" = "promoter")) %>%
  dplyr::rename(variant_expression = expression.x,
                control_expression = expression.y)
  
HAR_tag_seq_barcodes_masked_pD2_pval_comp <- HAR_tag_seq_barcodes_control_v_variant_masked_pD2 %>%
  dplyr::select(probe, barcode, mut_condition, promoter, rep, control_expression, variant_expression) %>%
  pivot_longer(cols = c(control_expression, variant_expression),
               names_to = "condition",
               values_to = "expression") %>%
  separate(col = condition,
           into = c("condition", "drop")) %>%
  dplyr::select(-c(drop)) %>%
  dplyr::distinct() %>%
  dplyr::group_by(probe, barcode) %>%
  t_test(expression ~ condition) %>%
  adjust_pvalue(p.col = "p", output.col = "p_adj", method = "BH")


write_tsv(HAR_tag_seq_barcodes_masked_pD2_pval_comp, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_pval_comp.txt")


# add variant info and save!
HAR_tag_seq_barcodes_masked_pD2_pval_comp_plus_vars <- left_join(HAR_tag_seq_barcodes_masked_pD2_pval_comp, HAR_caMPRA_variants_GIAB_masked, by = c("barcode" = "barcode"))

write_tsv(HAR_tag_seq_barcodes_masked_pD2_pval_comp_plus_vars, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_pval_comp_plus_vars.txt")

# 3572 variants (was 4001 prior to masking)
length(unique(HAR_tag_seq_barcodes_masked_pD2_pval_comp_plus_vars$var_id))

# combine control and variant information
HAR_tag_seq_barcodes_masked_pD2_comparison <- inner_join(
  HAR_tag_seq_barcodes_wo_variants_masked_pD2_stats, 
  HAR_tag_seq_barcodes_w_variants_masked_pD2_stats, 
  by = c("probe" = "probe")) %>%
  inner_join(., 
             HAR_tag_seq_barcodes_masked_pD2_pval_comp, 
             by = c("probe" = "probe", "barcode" = "barcode")) %>%
  dplyr::mutate(status = case_when(mean.y > mean.x & p_adj < 0.05 ~ "increased_expression",
                                   mean.x > mean.y & p_adj < 0.05 ~ "decreased_expression",
                                   TRUE ~ "not_significant"))

write_tsv(HAR_tag_seq_barcodes_masked_pD2_comparison, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison.txt")

# add variant information to stats + pval
HAR_tag_seq_barcodes_masked_pD2_comparison_plus_vars <- left_join(HAR_tag_seq_barcodes_masked_pD2_comparison, HAR_caMPRA_variants_GIAB_masked, by = c("barcode" = "barcode"))

write_tsv(HAR_tag_seq_barcodes_masked_pD2_comparison_plus_vars, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison_plus_vars.txt")

HAR_tag_seq_barcodes_masked_pD2_comparison_scatter_plot <- HAR_tag_seq_barcodes_masked_pD2_comparison %>%
  ggplot() +
  geom_point(aes(x = mean.x, y = mean.y, color = status, alpha = 0.25)) +
  geom_abline(slope = 1, intercept = 0) +
  theme_pubr()

HAR_tag_seq_barcodes_masked_pD2_comparison_bar_plot <- HAR_tag_seq_barcodes_masked_pD2_comparison %>%
  dplyr::group_by(status) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(status = case_when(status == "decreased_expression" ~ "decreased\nexpression",
                                   status == "increased_expression" ~ "increased\nexpression",
                                   status == "not_significant" ~ "minimal expression\nchange")) %>%
  ggplot() +
  geom_bar(aes(x = status, y = count, fill = status), stat = "identity") +
  theme_pubr()
  

# volcano plot of differences between variants and control samples? - note that the means are already in log2
HAR_tag_seq_barcodes_masked_pD2_comparison_volcano_plot <- HAR_tag_seq_barcodes_masked_pD2_comparison %>%
  ggplot() +
  geom_point(aes(x = mean.y - mean.x, y = -log10(p_adj), color = status, alpha = 0.25)) +
  xlab(label = "log2 FC\n(variant secquence vs control sequence)") +
  ylab(label = "-log10(p value)") +
  theme_pubr() +
  theme(legend.position = "none")

HAR_tag_seq_barcodes_masked_pD2_comparison_plots <- HAR_tag_seq_barcodes_masked_pD2_comparison_volcano_plot + HAR_tag_seq_barcodes_masked_pD2_comparison_bar_plot
ggsave(HAR_tag_seq_barcodes_masked_pD2_comparison_plots, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison_plots.png", height = 3, width = 9)
ggsave(HAR_tag_seq_barcodes_masked_pD2_comparison_plots, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison_plots_2.png", height = 5, width = 9)

ggsave(HAR_tag_seq_barcodes_masked_pD2_comparison_plots, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison_plots.pdf", height = 3, width = 9)
ggsave(HAR_tag_seq_barcodes_masked_pD2_comparison_plots, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison_plots_2.pdf", height = 5, width = 9)


rm(HAR_tag_seq_barcodes_masked_pD2_comparison_plots, HAR_tag_seq_barcodes_masked_pD2_comparison_scatter_plot, HAR_tag_seq_barcodes_masked_pD2_comparison_bar_plot)
```
```{r}
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD2_comparison_plots_2.png")
```
Note that there was an error in the previous version of this figure - the variant axis was way too high - was due to lack of transformation of that scale. Furthermore, the mean was taken for all replicates prior to log transformation.

From the preliminary analysis, it looks like the majority leads to non-significant change in expression, followed similar numbers of decreased and increased expression changes. A contributing factor to the non-sig bar is replicate variability - it looks like we may need higher transfection efficiency as well as earlier harvesting

09/15/22 update
updated code so that we mask GIAB variants, which cleans things up a bit (not a massive amount, but does improve things). Also, we are blinded from the GIAB variants, which is important for comparing control v variants

# pD3 Analysis
```{r, create a distribution baseline expression profiles for each HAR, eval = FALSE}
HAR_tag_seq_barcodes_wo_variants_masked_pD3 <- HAR_7x_mut_pD3_filtered_barcodes %>%
  filter(!barcode %in% HAR_caMPRA_variants_GIAB_masked$barcode) %>%
  select(HAR_7x_mut_pD3_TPM, 
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_5,
         probe) %>%
  dplyr::rename(
         HAR_7x_mut.pDonor3.rep_1 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut.pDonor3.rep_2 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut.pDonor3.rep_3 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut.pDonor3.rep_4 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut.pDonor3.rep_5 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_5
         ) %>%
  pivot_longer(-c(HAR_7x_mut_pD3_TPM, probe), names_to = "condition", values_to = "expression") %>%
  separate(col = condition, sep = "\\.", into = c("mut_condition", "promoter", "rep"))

HAR_tag_seq_barcodes_wo_variants_masked_pD3_stats <- HAR_tag_seq_barcodes_wo_variants_masked_pD3 %>% 
  dplyr::group_by(probe) %>%
  rstatix::get_summary_stats(expression)

write_tsv(HAR_tag_seq_barcodes_wo_variants_masked_pD3_stats, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_wo_variants_masked_pD3_stats.txt")

# with variants
HAR_tag_seq_barcodes_w_variants_masked_pD3 <- HAR_7x_mut_pD3_filtered_barcodes %>%
  filter(barcode %in% HAR_caMPRA_variants_GIAB_masked$barcode) %>%
  select(HAR_7x_mut_pD3_TPM, 
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_5,
         probe,
         barcode) %>%
  dplyr::rename(
         HAR_7x_mut.pDonor3.rep_1 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_1,
         HAR_7x_mut.pDonor3.rep_2 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_2,
         HAR_7x_mut.pDonor3.rep_3 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_3,
         HAR_7x_mut.pDonor3.rep_4 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_4,
         HAR_7x_mut.pDonor3.rep_5 = HAR_7x_mut_pD3_cDNA_tag_TPM_norm_log_5
         ) %>%
  pivot_longer(-c(HAR_7x_mut_pD3_TPM, probe, barcode), names_to = "condition", values_to = "expression") %>%
  separate(col = condition, sep = "\\.", into = c("mut_condition", "promoter", "rep"))

HAR_tag_seq_barcodes_w_variants_masked_pD3_stats <- HAR_tag_seq_barcodes_w_variants_masked_pD3 %>% 
  dplyr::group_by(probe, barcode) %>%
  rstatix::get_summary_stats(expression)

write_tsv(HAR_tag_seq_barcodes_w_variants_masked_pD3_stats, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_w_variants_masked_pD3_stats.txt")
```



```{r, HAR pD3 control v variant comparison, eval = FALSE}
# perform two sided paired t-tests to identify significantly different log2(mRNA/DNA) ratio between no variant sequences and variant sequences
HAR_tag_seq_barcodes_control_v_variant_masked_pD3 <- inner_join(
  HAR_tag_seq_barcodes_w_variants_masked_pD3, 
  HAR_tag_seq_barcodes_wo_variants_masked_pD3, 
  by = c("probe" = "probe", 
         "rep" = "rep",
         "mut_condition" = "mut_condition",
         "promoter" = "promoter")) %>%
  dplyr::rename(variant_expression = expression.x,
                control_expression = expression.y)


HAR_tag_seq_barcodes_masked_pD3_pval_comp <- HAR_tag_seq_barcodes_control_v_variant_masked_pD3 %>%
  dplyr::select(probe, barcode, mut_condition, promoter, rep, control_expression, variant_expression) %>%
  pivot_longer(cols = c(control_expression, variant_expression),
               names_to = "condition",
               values_to = "expression") %>%
  separate(col = condition,
           into = c("condition", "drop")) %>%
  dplyr::select(-c(drop)) %>%
  dplyr::distinct() %>%
  dplyr::group_by(probe, barcode) %>%
  t_test(expression ~ condition) %>%
  adjust_pvalue(p.col = "p", output.col = "p_adj", method = "BH")


write_tsv(HAR_tag_seq_barcodes_masked_pD3_pval_comp, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_pval_comp.txt")


# combine control and variant information
HAR_tag_seq_barcodes_masked_pD3_comparison <- inner_join(HAR_tag_seq_barcodes_wo_variants_masked_pD3_stats, HAR_tag_seq_barcodes_w_variants_masked_pD3_stats, by = c("probe" = "probe")) %>%
  inner_join(., HAR_tag_seq_barcodes_masked_pD3_pval_comp, by = c("probe" = "probe", "barcode" = "barcode")) %>%
  dplyr::mutate(status = case_when(mean.y > mean.x & p_adj < 0.05 ~ "increased_expression",
                                   mean.x > mean.y & p_adj < 0.05 ~ "decreased_expression",
                                   TRUE ~ "not_significant"))

# save information
write_tsv(HAR_tag_seq_barcodes_masked_pD3_comparison, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison")


HAR_tag_seq_barcodes_masked_pD3_comparison_scatter_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison %>%
  ggplot() +
  geom_point(aes(x = mean.x, y = mean.y, color = status, alpha = 0.25)) +
  geom_abline(slope = 1, intercept = 0) +
  theme_pubr()

HAR_tag_seq_barcodes_masked_pD3_comparison_bar_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison %>%
  dplyr::group_by(status) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(status = case_when(status == "decreased_expression" ~ "decreased\nexpression",
                                   status == "increased_expression" ~ "increased\nexpression",
                                   status == "not_significant" ~ "minimal expression\nchange")) %>%
  ggplot() +
  geom_bar(aes(x = status, y = count, fill = status), stat = "identity") +
  theme_pubr()
  

# volcano plot of differences between variants and control samples? - note that the means are already in log2
HAR_tag_seq_barcodes_masked_pD3_comparison_volcano_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison %>%
  ggplot() +
  geom_point(aes(x = mean.y - mean.x, y = -log10(p), color = status, alpha = 0.25)) +
  xlab(label = "log2 FC\n(variant secquence vs control sequence)") +
  ylab(label = "-log10(p value)") +
  theme_pubr() +
  theme(legend.position = "none")

HAR_tag_seq_barcodes_masked_pD3_comparison_plots <- HAR_tag_seq_barcodes_masked_pD3_comparison_volcano_plot + HAR_tag_seq_barcodes_masked_pD3_comparison_bar_plot
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_plots, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_plots.png", height = 3, width = 9)
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_plots, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_plots_2.png", height = 5, width = 9)

rm(HAR_tag_seq_barcodes_masked_pD3_comparison_plots, HAR_tag_seq_barcodes_masked_pD3_comparison_scatter_plot, HAR_tag_seq_barcodes_masked_pD3_comparison_bar_plot)

```
```{r}
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_plots_2.png")
```
pD3 results are very similar to pD2 results, which matches the finding from the sharpr paper that showned similar results despite the different promoters

# Check if variants are on same or different fragments
As we get variants that lead to changes in different directions, we need to check if they are:

1. in the same fragment - which we answered before - most 1 variant, but ~25% are 2 variants, 10% 3 variants. However, this does not affect the volcano plot as these are per probe basis. It will only affect associating individual variants to impact on TF binding
2. if we get different direction changes within the same HAR or if the direction of the change is largely dependent on which HAR is affected
3. How does the volcano plot look like in terms of variants (not as urgent)

```{r, check if variants are on same or different fragments, eval = FALSE}
# add variant information
HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info <- inner_join(x = HAR_tag_seq_barcodes_masked_pD3_comparison, y = variant_barcode_w_probe_info, by = c("barcode" = "barcode"))


# save information
write_tsv(HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info, file = "../Saved_Data/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info.txt")

# number of unique variants tested: 4532
str(unique(HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info$var_id))


# 7601 number of variants that make it through masking and filtering
str(unique(append(HAR_tag_seq_barcodes_masked_pD2_pval_comp_plus_vars$var_id, HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info$var_id)))

# see distribution by variant whether all barcodes associated with that barcode moves in the same direction. If not, see if there is a cut off that can be applied to filter out noise
# first focus on singletons
HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info_one_var_barcodes <- HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info %>%
  select(var_id, barcode, mean.x, mean.y, status) %>%
  group_by(barcode) %>%
  summarize(count = n()) %>%
  filter(count == 1)

# check how many of the singleton variants are discordant (have different barcodes indicating different signal)
HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_test <- HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info %>%
  select(var_id, probe, barcode, mean.x, mean.y, status) %>%
  # only select barcodes with one variant
  filter(barcode %in% HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info_one_var_barcodes$barcode) %>%
  group_by(probe, var_id, status) %>%
  summarize(count = n()) %>%
  pivot_wider(names_from = status, values_from = count, values_fill = 0) %>%
  dplyr::mutate(discordant = case_when(not_significant*increased_expression > 0 ~ "discordant",
                                       not_significant*decreased_expression > 0 ~ "discordant",
                                       decreased_expression*increased_expression > 0 ~ "discordant",
                                       TRUE ~ "concordant"))

# barcodes with concordance information
HAR_tag_seq_barcodes_masked_pD3_comparison_concordance <- HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info %>%
  # only select barcodes with one variant
  filter(barcode %in% HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info_one_var_barcodes$barcode) %>%
  full_join(x = .,
            y = HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_test,
            by = c("var_id" = "var_id", "probe" = "probe")) %>%
  select(barcode, discordant) %>%
  distinct() %>%
  full_join(x = HAR_tag_seq_barcodes_masked_pD3_comparison,
            y = .,
            by = c("barcode" = "barcode")) %>%
  dplyr::mutate(discordant = case_when(is.na(discordant) ~ "multi-variant",
                                       TRUE ~ discordant))

HAR_tag_seq_barcodes_masked_pD3_comparison_discordant <- HAR_tag_seq_barcodes_masked_pD3_comparison_concordance %>%
  filter(discordant == "discordant")

# volcano plot of differences between variants and control samples? - note that the means are already in log2
HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_volcano_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison_concordance %>%
  ggplot() +
  geom_point(aes(x = mean.y - mean.x, y = -log10(p_adj), color = status, alpha = 0.25)) +
  xlab(label = "log2 FC\n(variant secquence vs control sequence)") +
  ylab(label = "-log10(p value)") +
  geom_vline(xintercept = c(-0.2, 0.2)) +
  geom_hline(yintercept = 1.30103) +
  theme_pubr() +
  facet_wrap(~discordant)

HAR_tag_seq_barcodes_masked_pD3_comparison_mean_x_volcano_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison_concordance %>%
  ggplot() +
  geom_point(aes(x = mean.y - mean.x, y = -log10(p_adj), color = mean.x, alpha = 0.25)) +
  scale_color_distiller(palette = "Spectral") +
  xlab(label = "log2 FC\n(variant secquence vs control sequence)") +
  ylab(label = "-log10(p value adjusted)") +
  geom_vline(xintercept = c(-0.2, 0.2)) +
  geom_hline(yintercept = 1.30103) +
  theme_pubr() +
  facet_wrap(~discordant)

HAR_tag_seq_barcodes_masked_pD3_comparison_mean_y_volcano_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison_concordance %>%
  ggplot() +
  geom_point(aes(x = mean.y - mean.x, y = -log10(p_adj), color = mean.y, alpha = 0.25)) +
  scale_color_distiller(palette = "Spectral") +
  xlab(label = "log2 FC\n(variant secquence vs control sequence)") +
  ylab(label = "-log10(p value adjusted)") +
  geom_vline(xintercept = c(-0.2, 0.2)) +
  geom_hline(yintercept = 1.30103) +
  theme_pubr() +
  facet_wrap(~discordant)

# see by bar graph to see how much this impacts the analysis
HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_bar_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison_concordance %>%
  dplyr::group_by(status, discordant) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::mutate(status = case_when(status == "decreased_expression" ~ "decreased\nexpression",
                                   status == "increased_expression" ~ "increased\nexpression",
                                   status == "not_significant" ~ "minimal expression\nchange")) %>%
  ggplot() +
  geom_bar(aes(x = status, y = count, fill = discordant), stat = "identity") +
  theme_pubr()


# given the presence of the discordant variants, will remove them prior to downstream analysis

# per HAR/probe direction of variants - one variant probes
HAR_tag_seq_barcodes_masked_pD3_comparison_one_var_barcodes_per_probe_count <- HAR_tag_seq_barcodes_masked_pD3_comparison %>%
  # remove discordant barcodes
  filter(!(barcode %in% HAR_tag_seq_barcodes_masked_pD3_comparison_discordant$barcode)) %>%
  # filter for barcodes with just one variant
  filter(barcode %in% HAR_tag_seq_barcodes_masked_pD3_comparison_with_variant_info_one_var_barcodes$barcode) %>%
  dplyr::group_by(probe, status) %>%
  dplyr::summarise(count = n())

# see generate directional change classification for each probe
HAR_tag_seq_barcodes_masked_pD3_comparison_one_var_barcodes_classification <- HAR_tag_seq_barcodes_masked_pD3_comparison_one_var_barcodes_per_probe_count %>%
  pivot_wider(names_from = status, values_from = count, values_fill = 0) %>%
  dplyr::mutate(classification = case_when(decreased_expression > 0 & increased_expression > 0 ~ "both directions",
                                           decreased_expression > 0 & increased_expression == 0 ~ "decreased expression only",
                                           decreased_expression == 0 & increased_expression > 0 ~ "increased expression only",
                                           decreased_expression == 0 & increased_expression == 0 & not_significant > 0 ~ "no significant changes",
                                           TRUE ~ "ERROR"))

# visualize using bar graph
HAR_tag_seq_barcodes_masked_pD3_comparison_one_var_barcodes_classification_bar_plot <-
  HAR_tag_seq_barcodes_masked_pD3_comparison_one_var_barcodes_classification %>%
  dplyr::group_by(classification) %>%
  summarize(count = n()) %>%
  ggplot() +
  geom_bar(aes(x = classification, y = count, fill = classification), stat = "identity") +
  theme_pubr() +
  rotate_x_text(angle = 90)

# per HAR/probe direction of variants - single + multi variant probes
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification <-
  HAR_tag_seq_barcodes_masked_pD3_comparison %>%
  # remove discordant barcodes
  filter(!(barcode %in% HAR_tag_seq_barcodes_masked_pD3_comparison_discordant$barcode)) %>%
  dplyr::group_by(probe, status) %>%
  dplyr::summarise(count = n()) %>%
  pivot_wider(names_from = status, values_from = count, values_fill = 0) %>%
  dplyr::mutate(classification = case_when(decreased_expression > 0 & increased_expression > 0 ~ "both directions",
                                           decreased_expression > 0 & increased_expression == 0 ~ "decreased expression only",
                                           decreased_expression == 0 & increased_expression > 0 ~ "increased expression only",
                                           decreased_expression == 0 & increased_expression == 0 & not_significant > 0 ~ "no significant changes",
                                           TRUE ~ "ERROR"))

# visualize multi-variant classification of the probes
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_bar_plot <-
  HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification %>%
  dplyr::group_by(classification) %>%
  summarize(count = n()) %>%
  ggplot() +
  geom_bar(aes(x = classification, y = count, fill = classification), stat = "identity") +
  theme_pubr() +
  rotate_x_text(angle = 90)

# see generate directional change classification for each HAR (aggregate)
# HAR caMPRA - HAR overlap
# read in HAR regions
HAR_regions <- read_tsv("/Users/tshin/Dropbox (CA Walsh Lab)/Walsh Lab/Scripts/Projects/Reference/Regions/hg19_HARs_v2_merged.7.5.17.named.bed", col_names = c("chr", "start", "end", "HAR"))

# annotate caMPRA probes with HAR regions that overlap
HAR_caMPRA_probe_overlap <- genome_intersect(x = HAR_probe_names,
                                             y = HAR_regions,
                                             by = c("chr" = "chr", "start" = "start", "end" = "end"))

# group by HARs and perform classification
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR <- HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification %>%
  inner_join(x = .,
             y = HAR_caMPRA_probe_overlap,
             by = c("probe" = "probe")) %>%
  dplyr::select(probe, decreased_expression, increased_expression, not_significant, HAR) %>%
  group_by(HAR) %>%
  summarize(decreased_expression = sum(decreased_expression),
            increased_expression = sum(increased_expression),
            not_significant = sum(not_significant)) %>%
  dplyr::mutate(classification = case_when(decreased_expression > 0 & increased_expression > 0 ~ "both directions",
                                           decreased_expression > 0 & increased_expression == 0 ~ "decreased expression only",
                                           decreased_expression == 0 & increased_expression > 0 ~ "increased expression only",
                                           decreased_expression == 0 & increased_expression == 0 & not_significant > 0 ~ "no significant changes",
                                           TRUE ~ "ERROR"))

# visualize multi-variant classification of the probes
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_bar_plot <-
  HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR %>%
  dplyr::group_by(classification) %>%
  summarize(count = n()) %>%
  ggplot() +
  geom_bar(aes(x = classification, y = count, fill = classification), stat = "identity") +
  theme_pubr() +
  rotate_x_text(angle = 90)

# check to see distribution of classification with number of variants
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_probe_count_hist <- HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR %>%
  dplyr::mutate(probe_count = decreased_expression + increased_expression + not_significant) %>%
  ggplot() +
  geom_histogram(aes(x = probe_count, fill = classification), binwidth = 1) +
  facet_wrap(~classification) + 
  theme_pubr()

# grouped to calculate percentage of HARs that fall under different categories
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped <- HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR %>%
  ungroup() %>%
  dplyr::mutate(total = n()) %>%
  group_by(classification) %>%
  summarize(count = n(),
            percentage = count/total*100) %>%
  distinct() %>%
  ungroup() %>%
  dplyr::mutate(classification = factor(classification, levels = c("no significant changes", "increased expression only", "decreased expression only", "both directions")),
                percentage_rank = cumsum(percentage))

# plot in terms of percentages bar graph
HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped_bar_plot <- HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped %>%
  ggplot(.) +
  geom_col(aes(x = "classification of HARs", y = percentage, fill = classification)) +
  theme_pubr() +
  geom_text(data = HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped,
            aes(x = "classification of HARs", y = percentage_rank + 2, label = paste(count, "\n(", format(percentage, digits = 2, nsmall = 1), "%)")), vjust = 1.5, colour = "white")

# png
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_volcano_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_volcano_plot.png")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_bar_plot.png")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_probe_count_hist, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_probe_count_hist.png")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_bar_plot.png")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_bar_plot.png")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped_bar_plot.png", width = 3, height = 6)

# pdf
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_volcano_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_volcano_plot.pdf")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_bar_plot.pdf")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_probe_count_hist, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_probe_count_hist.pdf")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_bar_plot.pdf")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_bar_plot.pdf")
ggsave(HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped_bar_plot, filename = "../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped_bar_plot.pdf", width = 3, height = 6)
```
```{r}
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_volcano_plot.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_concordance_bar_plot.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_probe_count_hist.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_bar_plot.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_bar_plot.png")
knitr::include_graphics("../figures/Mutagenesis_Analysis/HAR_tag_seq_barcodes_masked_pD3_comparison_multi_var_barcodes_classification_by_HAR_grouped_bar_plot.png")
```


So it looks like around 4435 unique variants were tested with by the 4112 barcodes. It seems that occasionally there more than 1 barcode for each variant. However, there appear to be cases when we have only 1 of the barcodes being significant, while the others might not be.

There are 91 out of 1963 of the variants that are singletons (one variants per barcode) are discordant (~5%) 
However, this translates to 429 barcodes out of 2471 singleton barcodes that are affected (~18%).

There does not appear to be an easy way to remove the discordant barcodes. It is not clear if it is because there are hidden variants unaccounted for in the middle (possible) - however, we will remove discordant variant barcodes since they seem to be evenly distributed across different groups.

For future versions, incorporate ion torrent to capture all variants

In terms of how each caMPRA probe responds to variants, at least in terms of singleton probes (with just one variant), we get roughly an even split of just increased or just decreased expression probes. A smaller number of probes appear to have variants that go in both directions, which is interesting, but on the whole, the roughly even increased and decreased expression split is seems to be driven by individual elements that increased in one direction only. Note that given the very low level of mutagenesis, it is difficult to rule out that each element is bidirectionally sensitive.

To test this a bit, we can try including the multi-variant probes too. In this case, we see to be an elevated number of all categories (which makes sense). We still get roughly an equal number of increased/decreased only probes, but we also get a larger number of both direction probes. While the interpretation is a bit more tricky given there are more than 1 variant per probe, it seems that the probes are able to have changes in opposite directions with enough mutations.

The next step would be to aggregate this over the HARs.

Finally, we can perform in silico saturation mutagenesis on all the HARs (not as necessary but would be an interesting comparison).
At least we can look at deepsea based predictions for the specific variants to see if the results are at least similar prior to going ahead with the insilico mutagenesis

When grouped by HARs, it seems that we get an increase in the proportion of regions that are increased, decreased, or bi-directional changes. This seems to agree with the trend that most of the non-significant variant only HARs were those that only had 1 probe associated with it (which themselves are likely to only have 1 variant). It seems that even with the relatively low level of mutagenesis, half the HARs tested appear to have variants that change regulatory activity.


