---
title: "EXP_00010_Variant_TF_Analysis_Sample"
output: html_notebook
---

# Purpose
The aim of this notebook book is to examine how variants found in the cohorts impact TF binding sites (if any), and whether it segregates between affected and unaffected individuals in interesting ways.

# Import data
```{r, load libraries}
library(tidyverse)
library(ggpubr)
library(patchwork)
library(knitr)
library(readxl)
library(plyranges)
library(rstatix)
library(MotifDb)
library(motifbreakR)
library(BSgenome.Hsapiens.UCSC.hg19)
```

# TF analysis of rates

```{r, functions}
# function to format annovar output properly... (Annovar output is terrible...)
annovar_output_format <- function(annovar_bed, genome, test = FALSE){
  ## extract SNVs
  # given that annovar output for SNVs are 0 based while we BSgenomes is 1-based
  snvs <- annovar_bed %>%
    dplyr::filter(str_detect(ref, "^[:alpha:]+$") & 
                    str_detect(alt, "^[:alpha:]+$")) 
  
  # get granges for sequence just for ref/alt
  snvs_granges <- as_granges(snvs %>%
                               dplyr::mutate(start = end), seqnames = chr)
  
  snvs_corrected_formatted <- snvs %>%
    # add in ref from reference to check for differences
    dplyr::mutate(ref_sequence = BSgenome::getSeq(x = genome, snvs_granges) %>%
                    as.character(),
                  alt_sequence = alt
    ) %>%
    dplyr::rename(ref_origin = ref,
                  alt_origin = alt) %>%
    dplyr::rename(ref = ref_sequence,
                  alt = alt_sequence)
  
  # as snvs are fine in terms of sequence leave as is

  
  ## extract insertions
  ins <- annovar_bed %>%
    dplyr::filter(str_detect(alt, "^[:alpha:]+$") & 
                    (ref == "0" | ref == "-"))
    # for insertions, leave as 0 - based to extract sequence just before insertion
  # only precede if  ins is not empty
  if (plyr::empty(ins) == FALSE){
  # get granges for sequence just before ref/alt
  ins_pre_one_granges <- as_granges(ins %>%
                                      dplyr::mutate(end = start), seqnames = chr)
  
  ins_corrected <- ins %>%
    # add in sequence just before insertion
    dplyr::mutate(ref_sequence = BSgenome::getSeq(x = genome, ins_pre_one_granges) %>%
                    as.character(),
                  alt_sequence = paste0(
                    BSgenome::getSeq(x = genome, ins_pre_one_granges) %>%
                      as.character(), alt
                    )
                  ) %>%
    # fix start and end annotation (0-based)
    dplyr::mutate(start_fixed = start - 1)
  
  ins_corrected_formatted <- ins_corrected %>%
    dplyr::rename(ref_origin = ref,
                  alt_origin = alt) %>%
    dplyr::select(chr, start_fixed, end, ref_sequence, alt_sequence, everything(), -c(start)) %>%
    dplyr::rename(start = start_fixed,
                  ref = ref_sequence,
                  alt = alt_sequence)
  } else{
    ins_corrected <- tibble()
    ins_corrected_formatted <- tibble()
  }
  # extract deletions
  del <- annovar_bed %>%
    dplyr::filter(str_detect(ref, "^[:alpha:]+$") & 
                    (alt == "0"| alt == "-"))
    # for deletions, leave as 0 - based to extract sequence just before insertion
  # only continue if deletion tibble not empty
  if (plyr::empty(del) == FALSE){
  # get granges for sequence just before ref/alt
  del_pre_one_granges <- as_granges(del %>%
                                      dplyr::mutate(end = start), seqnames = chr)

  del_corrected <- del %>%
    # add in sequence just before deletion
    dplyr::mutate(alt_sequence = BSgenome::getSeq(x = genome, del_pre_one_granges) %>%
                    as.character(),
                  ref_sequence = paste0(
                    BSgenome::getSeq(x = genome, del_pre_one_granges) %>% as.character(), ref
                    )
                  ) %>%
    # fix start and end annotation (0-based)
    dplyr::mutate(start_fixed = start - 1)

   del_corrected_formatted <- del_corrected %>%
     dplyr::rename(ref_origin = ref,
                   alt_origin = alt) %>%
     dplyr::select(chr, start_fixed, end, ref_sequence, alt_sequence, everything(), -c(start)) %>%
     dplyr::rename(start = start_fixed,
                   ref = ref_sequence,
                   alt = alt_sequence)
  } else{
    del_corrected <- tibble()
    del_corrected_formatted <- tibble()
  }
  # combine into one tibble
  all_variants <- bind_rows(snvs_corrected_formatted,
                            ins_corrected_formatted,
                            del_corrected_formatted)
  
  # return output
  if (test == TRUE) {
    return(list(ins_corrected = ins_corrected, del_corrected = del_corrected, snvs_corrected = snvs_corrected_formatted))
  } else {
    return(all_variants)
  }
}

```



```{r, read in motifbreakR formatted bed files}
# note that motifbreakR requires a specific format (not really a bed file)
# read in original bed-like file and reformat
HAR_hg19_0.001_hom_bed <- read_tsv("/Users/tshin/Dropbox (CA Walsh Lab)/Walsh Lab/Scripts/Projects/Reference/Targeted_Sequencing/05_24_19_HAR_collapsed_Joint_call_final.AllCons.0.001.hom.chr.bed", col_names = c("chr", "start", "end", "ref", "alt", "zyg", "affected_status"))

CNE_hg19_0.001_hom_bed <- read_tsv("/Users/tshin/Dropbox (CA Walsh Lab)/Walsh Lab/Scripts/Projects/Reference/Targeted_Sequencing/05_24_19_NE_collapsed_Joint_call_final.AllCons.0.001.hom.chr.bed", col_names = c("chr", "start", "end", "ref", "alt", "zyg", "affected_status"))

VE_hg19_0.001_hom_bed <- read_tsv("/Users/tshin/Dropbox (CA Walsh Lab)/Walsh Lab/Scripts/Projects/Reference/Targeted_Sequencing/05_24_19_UCE_UCE_collapsed_Joint_call_final.AllCons.0.001.hom.chr.bed", col_names = c("chr", "start", "end", "ref", "alt", "zyg", "affected_status"))


HAR_hg19_0.001_hom_annovar_fixed <- annovar_output_format(annovar_bed = HAR_hg19_0.001_hom_bed, 
                              genome = BSgenome.Hsapiens.UCSC.hg19, 
                              test = FALSE) %>%
  # add motibreakR format
  dplyr::mutate(name = paste(chr, end, ref, alt, sep = ":"))

CNE_hg19_0.001_hom_annovar_fixed <- annovar_output_format(annovar_bed = CNE_hg19_0.001_hom_bed, 
                              genome = BSgenome.Hsapiens.UCSC.hg19, 
                              test = FALSE) %>%
  # add motibreakR format
  dplyr::mutate(name = paste(chr, end, ref, alt, sep = ":"))

VE_hg19_0.001_hom_annovar_fixed <- annovar_output_format(annovar_bed = VE_hg19_0.001_hom_bed, 
                              genome = BSgenome.Hsapiens.UCSC.hg19, 
                              test = FALSE) %>%
  # add motibreakR format
  dplyr::mutate(name = paste(chr, end, ref, alt, sep = ":"))


# count number of each variant (filter out weird indels)
HAR_hg19_0.001_hom_bed_counts <- HAR_hg19_0.001_hom_annovar_fixed %>%
  dplyr::group_by(name) %>%
  dplyr::summarise(count = n())

CNE_hg19_0.001_hom_bed_counts <- CNE_hg19_0.001_hom_annovar_fixed %>%
  dplyr::group_by(name) %>%
  dplyr::summarise(count = n())

VE_hg19_0.001_hom_bed_counts <- VE_hg19_0.001_hom_annovar_fixed %>%
  dplyr::group_by(name) %>%
  dplyr::summarise(count = n())

# remove variants that seem too overrepresented (indels that are highly prevalent - most likely artifacts)
HAR_hg19_0.001_hom_bed_counts_filtered_variants <- HAR_hg19_0.001_hom_bed_counts %>%
  # remove indels that look like artifacts
  dplyr::filter(count < 15)

CNE_hg19_0.001_hom_bed_counts_filtered_variants <- CNE_hg19_0.001_hom_bed_counts %>%
  # remove indels that look like artifacts
  dplyr::filter(count < 15)

VE_hg19_0.001_hom_bed_counts_filtered_variants <- VE_hg19_0.001_hom_bed_counts %>%
  # remove indels that look like artifacts
  dplyr::filter(count < 15)


# output fixed and filtered variant list
HAR_hg19_0.001_hom_bed_filtered <- HAR_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% HAR_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, name, score, strand) %>%
  distinct()

CNE_hg19_0.001_hom_bed_filtered <- CNE_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% CNE_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, name, score, strand) %>%
  distinct()

VE_hg19_0.001_hom_bed_filtered <- VE_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% VE_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, name, score, strand) %>%
  distinct()

write_tsv(HAR_hg19_0.001_hom_bed_filtered, file = "saved_data/HAR_hg19_0.001_hom_motifbreakR_formatted.bed", col_names = FALSE)
write_tsv(CNE_hg19_0.001_hom_bed_filtered, file = "saved_data/CNE_hg19_0.001_hom_motifbreakR_formatted.bed", col_names = FALSE)
write_tsv(VE_hg19_0.001_hom_bed_filtered, file = "saved_data/VE_hg19_0.001_hom_motifbreakR_formatted.bed", col_names = FALSE)
```


```{r, read in files for motifbreakR}
jaspar_2022_motifs <- query(MotifDb, andStrings=c("hsapiens"),
                orStrings=c("jaspar2022"))

# read in formatted file into motif breaker
HAR_hg19_0.001_variants <- variants.from.file(file = "saved_data/HAR_hg19_0.001_hom_motifbreakR_formatted.bed",
                                  search.genome = BSgenome.Hsapiens.UCSC.hg19,
                                  format = "bed")

CNE_hg19_0.001_variants <- variants.from.file(file = "saved_data/CNE_hg19_0.001_hom_motifbreakR_formatted.bed",
                                  search.genome = BSgenome.Hsapiens.UCSC.hg19,
                                  format = "bed")

VE_hg19_0.001_variants <- variants.from.file(file = "saved_data/VE_hg19_0.001_hom_motifbreakR_formatted.bed",
                                  search.genome = BSgenome.Hsapiens.UCSC.hg19,
                                  format = "bed")

```

```{r, motifbreakR run for HARs non consanguineous}
HAR_hg19_0.001_results <- motifbreakR(snpList = HAR_hg19_0.001_variants, filterp = TRUE,
                       pwmList = jaspar_2022_motifs,
                       threshold = 1e-4,
                       method = "log",
                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                       BPPARAM = BiocParallel::MulticoreParam())

saveRDS(HAR_hg19_0.001_results, file = "saved_data/HAR_hg19_0.001_results.RDS")
write_tsv(as.tibble(HAR_hg19_0.001_results), file = "saved_data/HAR_hg19_0.001_variants.motifbreakR.scan.txt")
```

```{r, motifbreakR run for CNEs non consanguineous}
CNE_hg19_0.001_results <- motifbreakR(snpList = CNE_hg19_0.001_variants, filterp = TRUE,
                       pwmList = jaspar_2022_motifs,
                       threshold = 1e-4,
                       method = "log",
                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                       BPPARAM = BiocParallel::MulticoreParam())

saveRDS(CNE_hg19_0.001_results, file = "saved_data/CNE_hg19_0.001_results.RDS")
write_tsv(as.tibble(CNE_hg19_0.001_results), file = "saved_data/CNE_hg19_0.001_variants.motifbreakR.scan.txt")
```

```{r, motifbreakR run for VEs non consanguineous}
VE_hg19_0.001_results <- motifbreakR(snpList = VE_hg19_0.001_variants, filterp = TRUE,
                       pwmList = jaspar_2022_motifs,
                       threshold = 1e-4,
                       method = "log",
                       bkg = c(A=0.25, C=0.25, G=0.25, T=0.25),
                       BPPARAM = BiocParallel::MulticoreParam())

saveRDS(VE_hg19_0.001_results, file = "saved_data/VE_hg19_0.001_results.RDS")
write_tsv(as.tibble(VE_hg19_0.001_results), file = "saved_data/VE_hg19_0.001_variants.motifbreakR.scan.txt")
```

```{r, visualize variants}
# add patient information
HAR_hg19_0.001_hom_filtered_patient_annotation <- HAR_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% HAR_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, ref, alt, name, score, strand, affected_status) %>%
  # remove SCZ samples
  dplyr::filter(affected_status != 3)


# annotate variants that are found in both affected and unaffected individuals
# for those that are found in both, indicate relative numbers (a naive/conservative approach because there are more unaffected)
HAR_hg19_0.001_hom_filtered_patient_annotation_counts <- HAR_hg19_0.001_hom_filtered_patient_annotation %>%
  dplyr::group_by(name, affected_status) %>%
  dplyr::summarise(count = dplyr::n()) %>%
  pivot_wider(names_from = affected_status, names_prefix = "affected_stat_", values_from = count, values_fill = 0) %>%
  dplyr::mutate(affected_only = case_when(affected_stat_2 > 0 & affected_stat_1 == 0 ~ "affected_only",
                                          TRUE ~ "not_affected_only"),
                affected_enrichment = affected_stat_2 - affected_stat_1) %>%
  dplyr::select(name, affected_only, affected_enrichment) %>%
  dplyr::mutate(affected_enrichment_status = case_when(affected_enrichment > 0 ~ "affected_enriched",
                                                       TRUE ~ "not_affected_enriched"))


HAR_results_db <- HAR_hg19_0.001_results %>%
  as_tibble() %>%
  left_join(x = .,
            y = HAR_hg19_0.001_hom_filtered_patient_annotation %>%
              dplyr::mutate(name_adjusted = case_when(
                str_length(ref) > 1 | str_length(alt) > 1 ~ paste0(chr, ":", start + 1,"-", end, ":", ref, ":", alt),
                TRUE ~ name)),
            by = c("SNP_id" = "name_adjusted")) %>%
  # remove snps that were filtered before (like the SCZ snps)
  dplyr::filter(name %in% HAR_hg19_0.001_hom_filtered_patient_annotation$name) %>%
  left_join(x = .,
            y = HAR_hg19_0.001_hom_filtered_patient_annotation_counts,
            by = c("name" = "name"))


# need to visualize effect, as well as the fact that some variants have many TF overlaps, and some variants are in both affected and unaffected

# visualize HAR TF results
HAR_results_db_plot <- HAR_hg19_0.001_results %>%
  ggplot() +
  geom_density(aes(x = alleleEffectSize, color = as.factor(affected_enrichment))) +
  facet_wrap(~as.factor(affected_enrichment) + affected_only + effect, scales = "free")

HAR_results_db_plot_2 <- HAR_results_db %>%
  ggplot() +
  geom_point(aes(x = geneSymbol, y = alleleEffectSize, color = as.factor(affected_only)))

# count total number of unique variants (need to count non-motif containing elements too)
HAR_results_variant_counts <- HAR_hg19_0.001_hom_filtered_patient_annotation_counts %>%
  dplyr::select(name, affected_enrichment_status) %>%
  distinct() %>%
  dplyr::group_by(affected_enrichment_status) %>%
  dplyr::summarise(total_count =  dplyr::n())


# for each TF, make aggregation/count
HAR_results_db_TF_counts <- HAR_results_db %>%
  #dplyr::filter(effect == "strong") %>%
  dplyr::group_by(geneSymbol, affected_enrichment_status) %>%
  dplyr::summarise(count =  dplyr::n())


# aggregate TF count
HAR_results_db_aggregate_counts <- HAR_results_db %>%
  #dplyr::filter(effect == "strong") %>%
  dplyr::group_by(affected_enrichment_status) %>%
  dplyr::summarise(count =  dplyr::n())



# aggregate TF count (max 1 TF per variant)
HAR_results_db_aggregate_counts <- HAR_results_db %>%
  #dplyr::filter(effect == "strong") %>%
  dplyr::group_by(affected_enrichment_status, name) %>%
  dplyr::summarise(count =  dplyr::n())  %>%
  dplyr::group_by(affected_enrichment_status) %>%
  dplyr::summarise(count = dplyr::n())


# add totals
HAR_TF_counts_totals <- HAR_results_db_aggregate_counts %>%
  # complete missing categories (if there are no variants, then fill with 0)
  dplyr::ungroup() %>%
  tidyr::complete(affected_enrichment_status, fill = list(count = 0)) %>%
  left_join(.,
            y = HAR_results_variant_counts,
            by = c("affected_enrichment_status" = "affected_enrichment_status")) %>%
  dplyr::mutate(proportion = count/total_count) %>%
  pivot_wider(names_from = affected_enrichment_status,
              values_from = c(proportion, count, total_count),
              values_fill = 0)

# test hypergeometric test
HAR_TF_stat_test <- HAR_TF_counts_totals %>%
  ungroup() %>%
  rowwise() %>%
  dplyr::mutate(
    enrichment_score = log2(
      (count_affected_enriched / total_count_affected_enriched) /
        ((count_not_affected_enriched + count_affected_enriched) /
           (total_count_affected_enriched + total_count_not_affected_enriched)))
  ) %>%
  dplyr::mutate(
    p_value = phyper(
      q = count_affected_enriched - 1,
      m = total_count_affected_enriched,
      n = total_count_not_affected_enriched,
      k = count_affected_enriched + count_not_affected_enriched,
      lower.tail = FALSE
    )
  ) %>%
    ungroup() %>%
    adjust_pvalue(p = "p_value", method = "BH")

# quick visualization - not great because the low counts make it so they overlap...
HAR_TF_prop_plot <- HAR_TF_stat_test %>%
  arrange(enrichment_score) %>%
  #dplyr::slice_tail(n = 10) %>%
  dplyr::mutate(geneSymbol = factor(geneSymbol, levels = geneSymbol)) %>%
  ggplot() +
  geom_col(aes(x = enrichment_score, y = geneSymbol, color = p_value.adj < 0.05))

```


```{r, general function to access enrichment of TFBS impacted by variants}
# 3 different analyses
# 1. TF enrichment based on variants that are enriched in cases v controls (nominal excess, not accounting for number affected/unaffected) - must use only for # case varints >= # control variants
# 2. TF enrichment based on affected only variants and not affected only
# 3. TF enrichment based on affected only variants and unaffected only variants (will lose many variants but more stringent?)
# 4. Combined TF enrichment based on counts of strong/weak TF hits across all TFs and aggregating them

# 1. TF enrichment based on variants that are enriched in cases v controls (nominal excess, not accounting for number affected/unaffected) - must use only for # case varints >= # control variants
TF_enrichment_test <- function(motifbreakR_results, variants_with_affected_status){
  # input expectations
  # expects regular motifbreakR results with SNP_id metacolumn
  # expects variants with affected status to info to have columns: name (snp id info), 
  # affected_status (whether the variant is from affected or not affected individual
  # note that filtering of variants should be done prior to inputting into function
  
  # note that the function is assessing whether certain TFBS are disproportionately affected by SNVs 
  # from affected individuals. Each allele is classified as being enriched (nominally) 
  # for patient variants vs unaffected variants. This numerical excess does not take into consideration
  # total number of patient or unaffected individual variants (so need to be careful about imbalance).
  # This function is meant to be used when we have equal or less affected patient variants vs unaffected variants.
  
  # count number of each variants in affected and unaffected individuals. CLassify enrichment
  variants_with_affected_status_counts <- variants_with_affected_status %>%
  dplyr::group_by(name, affected_status) %>%
  dplyr::summarise(count = dplyr::n()) %>%
  pivot_wider(names_from = affected_status, names_prefix = "affected_stat_", values_from = count, values_fill = 0) %>%
  dplyr::mutate(affected_only = case_when(affected_stat_2 > 0 & affected_stat_1 == 0 ~ "affected_only",
                                          TRUE ~ "not_affected_only"),
                affected_enrichment = affected_stat_2 - affected_stat_1) %>%
  dplyr::select(name, affected_only, affected_enrichment) %>%
  dplyr::mutate(affected_enrichment_status = case_when(affected_enrichment > 0 ~ "affected_enriched",
                                                       TRUE ~ "not_affected_enriched"))
  
  # annotate motifbreakR results
  motifbreakR_results_annotated <- motifbreakR_results %>%
    as_tibble() %>%
    left_join(x = .,
              y = variants_with_affected_status %>%
                dplyr::mutate(name_adjusted = case_when(
                  str_length(ref) > 1 | str_length(alt) > 1 ~ paste0(chr, ":", start + 1,"-", end, ":", ref, ":", alt),
                  TRUE ~ name)),
              by = c("SNP_id" = "name_adjusted")) %>%
    # remove snps that are not annotated in the patient variant list 
    # (useful motifbreakR results was general and needs filtering)
    dplyr::filter(name %in% variants_with_affected_status$name) %>%
    left_join(x = .,
              y = variants_with_affected_status_counts,
              by = c("name" = "name"))
  

  # count total number of unique variants (need to count non-motif containing elements too)
  variants_total_counts <- variants_with_affected_status_counts %>%
    dplyr::select(name, affected_enrichment_status) %>%
    distinct() %>%
    dplyr::group_by(affected_enrichment_status) %>%
    dplyr::summarise(total_count =  dplyr::n())
  
  
  # for each TF, make aggregation/count
  motifbreakR_TF_counts <- motifbreakR_results_annotated %>%
    dplyr::group_by(geneSymbol, affected_enrichment_status) %>%
    dplyr::summarise(count =  dplyr::n())
  
  # add totals
  motifbreakR_TF_counts_w_totals <- motifbreakR_TF_counts %>%
    # complete missing categories (if there are no variants, then fill with 0)
    dplyr::ungroup() %>%
    tidyr::complete(geneSymbol, affected_enrichment_status, fill = list(count = 0)) %>%
    left_join(.,
              y = variants_total_counts,
              by = c("affected_enrichment_status" = "affected_enrichment_status")) %>%
    dplyr::mutate(proportion = count/total_count) %>%
    pivot_wider(names_from = affected_enrichment_status,
                values_from = c(proportion, count, total_count),
                values_fill = 0)
  
  # perform hypergeometric test for enrichment - BH adjusted p_values
  motifbreakR_TF_enrichment_test <- motifbreakR_TF_counts_w_totals %>%
    ungroup() %>%
    rowwise() %>%
    dplyr::mutate(
      enrichment_score = log2(
        (count_affected_enriched / total_count_affected_enriched) / 
          ((count_not_affected_enriched + count_affected_enriched) / 
             (total_count_affected_enriched + total_count_not_affected_enriched)))
      ) %>%
    dplyr::mutate(
      p_value = phyper(
        q = count_affected_enriched - 1,
        m = total_count_affected_enriched,
        n = total_count_not_affected_enriched,
        k = count_affected_enriched + count_not_affected_enriched,
        lower.tail = FALSE
      )
    ) %>%
    ungroup() %>%
    adjust_pvalue(p = "p_value", method = "BH")
  return(motifbreakR_TF_enrichment_test)
}


# 2. TF enrichment based on affected only variants and not affected only
TF_enrichment_affected_only_test <- function(motifbreakR_results, variants_with_affected_status){
  # input expectations
  # expects regular motifbreakR results with SNP_id metacolumn
  # expects variants with affected status to info to have columns: name (snp id info), 
  # affected_status (whether the variant is from affected or not affected individual
  # note that filtering of variants should be done prior to inputting into function
  
  # note that the function is assessing whether certain TFBS are disproportionately affected by SNVs 
  # from affected individuals. Each allele is classified as being enriched (nominally) 
  # for patient variants vs unaffected variants. This numerical excess does not take into consideration
  # total number of patient or unaffected individual variants (so need to be careful about imbalance).
  # This function is meant to be used when we have equal or less affected patient variants vs unaffected variants.
  
  # count number of each variants in affected and unaffected individuals. CLassify enrichment
  variants_with_affected_status_counts <- variants_with_affected_status %>%
  dplyr::group_by(name, affected_status) %>%
  dplyr::summarise(count = dplyr::n()) %>%
  pivot_wider(names_from = affected_status, names_prefix = "affected_stat_", values_from = count, values_fill = 0) %>%
  dplyr::mutate(affected_only = case_when(affected_stat_2 > 0 & affected_stat_1 == 0 ~ "affected_only",
                                          TRUE ~ "not_affected_only"),
                affected_enrichment = affected_stat_2 - affected_stat_1) %>%
  dplyr::select(name, affected_only, affected_enrichment) %>%
  dplyr::mutate(affected_enrichment_status = case_when(affected_enrichment > 0 ~ "affected_enriched",
                                                       TRUE ~ "not_affected_enriched"))
  
  # annotate motifbreakR results
  motifbreakR_results_annotated <- motifbreakR_results %>%
    as_tibble() %>%
    left_join(x = .,
              y = variants_with_affected_status %>%
                dplyr::mutate(name_adjusted = case_when(
                  str_length(ref) > 1 | str_length(alt) > 1 ~ paste0(chr, ":", start + 1,"-", end, ":", ref, ":", alt),
                  TRUE ~ name)),
              by = c("SNP_id" = "name_adjusted")) %>%
    # remove snps that are not annotated in the patient variant list 
    # (useful motifbreakR results was general and needs filtering)
    dplyr::filter(name %in% variants_with_affected_status$name) %>%
    left_join(x = .,
              y = variants_with_affected_status_counts,
              by = c("name" = "name"))
  

  # count total number of unique variants (need to count non-motif containing elements too)
  variants_total_counts <- variants_with_affected_status_counts %>%
    dplyr::select(name, affected_only) %>%
    distinct() %>%
    dplyr::group_by(affected_only) %>%
    dplyr::summarise(total_count =  dplyr::n())
  
  
  # for each TF, make aggregation/count
  motifbreakR_TF_counts <- motifbreakR_results_annotated %>%
    dplyr::group_by(geneSymbol, affected_only) %>%
    dplyr::summarise(count =  dplyr::n())
  
  # add totals
  motifbreakR_TF_counts_w_totals <- motifbreakR_TF_counts %>%
    # complete missing categories (if there are no variants, then fill with 0)
    dplyr::ungroup() %>%
    tidyr::complete(geneSymbol, affected_only, fill = list(count = 0)) %>%
    left_join(.,
              y = variants_total_counts,
              by = c("affected_only" = "affected_only")) %>%
    dplyr::mutate(proportion = count/total_count) %>%
    pivot_wider(names_from = affected_only,
                values_from = c(proportion, count, total_count),
                values_fill = 0)
  
  # perform hypergeometric test for enrichment - BH adjusted p_values
  motifbreakR_TF_enrichment_test <- motifbreakR_TF_counts_w_totals %>%
    dplyr::ungroup() %>%
    dplyr::rowwise() %>%
    dplyr::mutate(
      enrichment_score = log2(
        (count_affected_only / total_count_affected_only) / 
          ((count_not_affected_only + count_affected_only) /
             (total_count_affected_only + total_count_not_affected_only)))
    ) %>%
    dplyr::mutate(
      p_value = phyper(
        q = count_affected_only - 1,
        m = total_count_affected_only,
        n = total_count_not_affected_only,
        k = count_affected_only + count_not_affected_only,
        lower.tail = FALSE
      )
    ) %>%
    ungroup() %>%
    adjust_pvalue(p = "p_value", method = "BH")
  return(motifbreakR_TF_enrichment_test)
}


# 3. TF enrichment based on affected only variants and unaffected only variants (will lose many variants but more stringent?)
TF_enrichment_affected_unaffected_only_test <- function(motifbreakR_results, variants_with_affected_status){
  # input expectations
  # expects regular motifbreakR results with SNP_id metacolumn
  # expects variants with affected status to info to have columns: name (snp id info), 
  # affected_status (whether the variant is from affected or not affected individual
  # note that filtering of variants should be done prior to inputting into function
  
  # note that the function is assessing whether certain TFBS are disproportionately affected by SNVs 
  # from affected individuals. Each allele is classified as being enriched (nominally) 
  # for patient variants vs unaffected variants. This numerical excess does not take into consideration
  # total number of patient or unaffected individual variants (so need to be careful about imbalance).
  # This function is meant to be used when we have equal or less affected patient variants vs unaffected variants.
  
  # count number of each variants in affected and unaffected individuals
  variants_with_affected_status_counts <- variants_with_affected_status %>%
    dplyr::group_by(name, affected_status) %>%
    dplyr::summarise(count = dplyr::n()) %>%
    pivot_wider(names_from = affected_status, names_prefix = "affected_stat_", values_from = count, values_fill = 0) %>%
    dplyr::mutate(affected_only = case_when(affected_stat_2 > 0 & affected_stat_1 == 0 ~ "affected_only",
                                            affected_stat_1 > 0 & affected_stat_2 == 0 ~ "unaffected_only",
                                            TRUE ~ "shared"),
                  affected_enrichment = affected_stat_2 - affected_stat_1) %>%
    # remove variants that are in both affected and unaffected individuals
    dplyr::filter(affected_only != "shared") %>%
    dplyr::select(name, affected_only, affected_enrichment) %>%
    dplyr::mutate(affected_enrichment_status = case_when(affected_enrichment > 0 ~ "affected_enriched",
                                                         TRUE ~ "not_affected_enriched"))
  
  # annotate motifbreakR results
  motifbreakR_results_annotated <- motifbreakR_results %>%
    as_tibble() %>%
    left_join(x = .,
              y = variants_with_affected_status %>%
                dplyr::mutate(name_adjusted = case_when(
                  str_length(ref) > 1 | str_length(alt) > 1 ~ paste0(chr, ":", start + 1,"-", end, ":", ref, ":", alt),
                  TRUE ~ name)),
              by = c("SNP_id" = "name_adjusted")) %>%
    # remove snps that are not annotated in the patient variant list 
    # (useful motifbreakR results was general and needs filtering)
    dplyr::filter(name %in% variants_with_affected_status$name) %>%
    left_join(x = .,
              y = variants_with_affected_status_counts,
              by = c("name" = "name"))
  

  # count total number of unique variants (need to count non-motif containing elements too)
  variants_total_counts <- variants_with_affected_status_counts %>%
    dplyr::select(name, affected_only) %>%
    distinct() %>%
    dplyr::group_by(affected_only) %>%
    dplyr::summarise(total_count =  dplyr::n())
  
  
  # for each TF, make aggregation/count
  motifbreakR_TF_counts <- motifbreakR_results_annotated %>%
    dplyr::group_by(geneSymbol, affected_only) %>%
    dplyr::summarise(count =  dplyr::n())
  
  # add totals
  motifbreakR_TF_counts_w_totals <- motifbreakR_TF_counts %>%
    # complete missing categories (if there are no variants, then fill with 0)
    dplyr::ungroup() %>%
    tidyr::complete(geneSymbol, affected_only, fill = list(count = 0)) %>%
    left_join(.,
              y = variants_total_counts,
              by = c("affected_only" = "affected_only")) %>%
    dplyr::mutate(proportion = count/total_count) %>%
    pivot_wider(names_from = affected_only,
                values_from = c(proportion, count, total_count),
                values_fill = 0)
  
  # perform hypergeometric test for enrichment - BH adjusted p_values
  motifbreakR_TF_enrichment_test <- motifbreakR_TF_counts_w_totals %>%
    ungroup() %>%
    rowwise() %>%
    dplyr::mutate(
      enrichment_score = log2(
        (count_affected_only / total_count_affected_only) / 
          ((count_unaffected_only + count_affected_only) / 
             (total_count_affected_only + total_count_unaffected_only)))
      ) %>%
    dplyr::mutate(
      p_value = phyper(
        q = count_affected_only - 1,
        m = total_count_affected_only,
        n = total_count_unaffected_only,
        k = count_affected_only + count_unaffected_only,
        lower.tail = FALSE
      )
    ) %>%
    ungroup() %>%
    adjust_pvalue(p = "p_value", method = "BH")
  return(motifbreakR_TF_enrichment_test)
}


# 4. Combined TF enrichment based on counts of strong/weak TF hits across all TFs and aggregating them
TF_aggregate_enrichment_test <- function(motifbreakR_results, variants_with_affected_status){
  # input expectations
  # expects regular motifbreakR results with SNP_id metacolumn
  # expects variants with affected status to info to have columns: name (snp id info), 
  # affected_status (whether the variant is from affected or not affected individual
  # note that filtering of variants should be done prior to inputting into function
  
  # note that the function is assessing whether certain TFBS are disproportionately affected by SNVs 
  # from affected individuals. Each allele is classified as being enriched (nominally) 
  # for patient variants vs unaffected variants. This numerical excess does not take into consideration
  # total number of patient or unaffected individual variants (so need to be careful about imbalance).
  # This function is meant to be used when we have equal or less affected patient variants vs unaffected variants.
  
  # count number of each variants in affected and unaffected individuals. CLassify enrichment
  variants_with_affected_status_counts <- variants_with_affected_status %>%
  dplyr::group_by(name, affected_status) %>%
  dplyr::summarise(count = dplyr::n()) %>%
  pivot_wider(names_from = affected_status, names_prefix = "affected_stat_", values_from = count, values_fill = 0) %>%
  dplyr::mutate(affected_only = case_when(affected_stat_2 > 0 & affected_stat_1 == 0 ~ "affected_only",
                                          TRUE ~ "not_affected_only"),
                affected_enrichment = affected_stat_2 - affected_stat_1) %>%
  dplyr::select(name, affected_only, affected_enrichment) %>%
  dplyr::mutate(affected_enrichment_status = case_when(affected_enrichment > 0 ~ "affected_enriched",
                                                       TRUE ~ "not_affected_enriched"))
  
  # annotate motifbreakR results
  motifbreakR_results_annotated <- motifbreakR_results %>%
    as_tibble() %>%
    left_join(x = .,
              y = variants_with_affected_status %>%
                dplyr::mutate(name_adjusted = case_when(
                  str_length(ref) > 1 | str_length(alt) > 1 ~ paste0(chr, ":", start + 1,"-", end, ":", ref, ":", alt),
                  TRUE ~ name)),
              by = c("SNP_id" = "name_adjusted")) %>%
    # remove snps that are not annotated in the patient variant list 
    # (useful motifbreakR results was general and needs filtering)
    dplyr::filter(name %in% variants_with_affected_status$name) %>%
    left_join(x = .,
              y = variants_with_affected_status_counts,
              by = c("name" = "name"))
  

  # count total number of unique variants (need to count non-motif containing elements too)
  variants_total_counts <- variants_with_affected_status_counts %>%
    dplyr::select(name, affected_enrichment_status) %>%
    distinct() %>%
    dplyr::group_by(affected_enrichment_status) %>%
    dplyr::summarise(total_count =  dplyr::n())
  
  
  # for each TF, make aggregation/count (only 1 TF)
  motifbreakR_TF_aggregate_counts <- motifbreakR_results_annotated %>%
    dplyr::group_by(affected_enrichment_status, name) %>%
    dplyr::summarise(count =  dplyr::n()) %>%
    dplyr::group_by(affected_enrichment_status) %>%
    dplyr::summarise(count = dplyr::n())
  
  # add totals
  motifbreakR_TF_counts_w_totals <- motifbreakR_TF_aggregate_counts %>%
    # complete missing categories (if there are no variants, then fill with 0)
    dplyr::ungroup() %>%
    tidyr::complete(affected_enrichment_status, fill = list(count = 0)) %>%
    left_join(.,
              y = variants_total_counts,
              by = c("affected_enrichment_status" = "affected_enrichment_status")) %>%
    dplyr::mutate(proportion = count/total_count) %>%
    pivot_wider(names_from = affected_enrichment_status,
                values_from = c(proportion, count, total_count),
                values_fill = 0)
  
  # perform hypergeometric test for enrichment - BH adjusted p_values
  motifbreakR_TF_enrichment_test <- motifbreakR_TF_counts_w_totals %>%
    ungroup() %>%
    rowwise() %>%
    dplyr::mutate(
      enrichment_score = log2(
        (count_affected_enriched / total_count_affected_enriched) / 
          ((count_not_affected_enriched + count_affected_enriched) / 
             (total_count_affected_enriched + total_count_not_affected_enriched)))
      ) %>%
    dplyr::mutate(
      p_value = phyper(
        q = count_affected_enriched - 1,
        m = total_count_affected_enriched,
        n = total_count_not_affected_enriched,
        k = count_affected_enriched + count_not_affected_enriched,
        lower.tail = FALSE
      )
    ) %>%
    ungroup() %>%
    adjust_pvalue(p = "p_value", method = "BH")
  return(motifbreakR_TF_enrichment_test)
}
```

```{r, list of patient variants}
# add patient information
HAR_hg19_0.001_hom_filtered_patient_annotation <- HAR_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% HAR_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, ref, alt, name, score, strand, affected_status) %>%
  # remove SCZ samples
  dplyr::filter(affected_status != 3)


# add patient information
CNE_hg19_0.001_hom_filtered_patient_annotation <- CNE_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% CNE_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, ref, alt, name, score, strand, affected_status) %>%
  # remove SCZ samples
  dplyr::filter(affected_status != 3)

# add patient information
VE_hg19_0.001_hom_filtered_patient_annotation <- VE_hg19_0.001_hom_annovar_fixed %>%
  dplyr::filter(name %in% VE_hg19_0.001_hom_bed_counts_filtered_variants$name) %>%
  dplyr::mutate(score = 0,
                strand = "*") %>%
  dplyr::select(chr, start, end, ref, alt, name, score, strand, affected_status) %>%
  # remove SCZ samples
  dplyr::filter(affected_status != 3)

```


```{r, test variant classification}
# HARs
HAR_NIMH_AGRE_0.001_TF_enrichment <- TF_enrichment_test(
  motifbreakR_results = HAR_hg19_0.001_results, 
  variants_with_affected_status = HAR_hg19_0.001_hom_filtered_patient_annotation)
```


```{r, set up motifbreakR for all variants}
# non-consanguineous variants
# HARs
# already set up for HARs
HAR_NIMH_AGRE_0.001_TF_enrichment <- TF_enrichment_test(
  motifbreakR_results = HAR_hg19_0.001_results, 
  variants_with_affected_status = HAR_hg19_0.001_hom_filtered_patient_annotation)

CNE_NIMH_AGRE_0.001_TF_enrichment <- TF_enrichment_test(
  motifbreakR_results = CNE_hg19_0.001_results, 
  variants_with_affected_status = CNE_hg19_0.001_hom_filtered_patient_annotation)

VE_NIMH_AGRE_0.001_TF_enrichment <- TF_enrichment_test(
  motifbreakR_results = VE_hg19_0.001_results, 
  variants_with_affected_status = VE_hg19_0.001_hom_filtered_patient_annotation)

# combined
HAR_VE_NIMH_AGRE_0.001_TF_enrichment <- TF_enrichment_test(
  motifbreakR_results = bind_ranges(HAR_hg19_0.001_results, 
                                    VE_hg19_0.001_results), 
  variants_with_affected_status = bind_rows(HAR_hg19_0.001_hom_filtered_patient_annotation,
                                            VE_hg19_0.001_hom_filtered_patient_annotation))


```






